### PR TITLE
Improve config default lookup and `--help` message

### PR DESCRIPTION
**Pull request by tiran**
Introduce a `click.Option` subclass that renders help messages with defaults from the context's `default_map` and shows a hint to the config setting name. This simplifies the code and avoids code duplication.

The implementation can later be extended to show dynamic defaults from the user's config file instead of static defaults from the default config. The new `ConfigOption` can lookup defaults from nested settings. There is no need to flatten the configuration any more.

## Simple example `ilab model serve`

Code:
```python
@click.option(
    "--model-path",
    type=click.Path(),
    cls=ConfigOption,
    help="Path to the model used during generation.",
)
```

Output:
```
Options:
  --model-path PATH           Path to the model used during generation.
                              [default: models/merlinite-7b-lab-Q4_K_M.gguf;
                              config: 'serve.model_path']
```

## Complex example with nested lookup in `ilab model evaluate`

```python
@click.option(
    "--judge-model",
    type=click.STRING,
    cls=clickext.ConfigOption,
    config_sections="mt_bench",
    help="Model to be used as a judge for running mt_bench or mt_bench_branch - can be a local path or the name of a Hugging Face repository",
)
```

Output:
```
$ ilab model evaluate --help
Options:
...
  --judge-model TEXT              Model to be used as a judge for running
                                  mt_bench or mt_bench_branch - can be a local
                                  path or the name of a Hugging Face
                                  repository  [default: prometheus-
                                  eval/prometheus-8x7b-v2.0; config:
                                  'evaluate.mt_bench.judge_model']
...
```

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [X] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

**TODO:**
- [x] Add `ConfigOption` to all subcommands
- [x] Extend `ConfigOption` to support nested defaults, e.g. `llama_cpp.gpu_layers`

### CODE REVIEW

File: src/instructlab/model/serve.py [hunk 1]
```diff
@@ -20,17 +19,23 @@
...
-    default=config.DEFAULT_MODEL_PATH,
-    show_default=True,
+    cls=clickext.ConfigOption,
+    type=click.Path(path_type=pathlib.Path),
+    required=True,  # default from config
     help="Path to the model used during generation.",
 )
 @click.option(
     "--gpu-layers",
+    # cls=ConfigOption,
```
Comment (leseb) [inline src/instructlab/model/serve.py:29]: left over?

File: src/instructlab/model/serve.py [hunk 2]
```diff
@@ -20,17 +19,23 @@
 @click.command()
 @click.option(
     "--model-path",
-    type=click.Path(),
-    default=config.DEFAULT_MODEL_PATH,
-    show_default=True,
+    cls=clickext.ConfigOption,
+    type=click.Path(path_type=pathlib.Path),
+    required=True,  # default from config
```
Comment (leseb) [inline src/instructlab/model/serve.py:24]: Why is the comment needed?

Comment (tiran) [reply src/instructlab/model/serve.py:24]: For humans. Or can you tell from the code that the default is coming from a mandatory field in `config.yaml`?

File: src/instructlab/model/serve.py (hunk 1 above)
Comment (tiran) [reply src/instructlab/model/serve.py:29]: No, not implemented yet. The default comes from `llama_cpp.gpu_layers`, not `gpu_layers`.

File: src/instructlab/model/serve.py (hunk 2 above)
Comment (leseb) [reply src/instructlab/model/serve.py:24]: I'm human, I guess I wasn't sure how to read it :)

**Comment by tiran** [comment]
@cdoern Please take a look. I have extended click, so we now look up defaults in nested default_map entries and report the config location in `--help`:

```
$ ilab model train --help
...
Options:
  --data-path PATH                Base directory where data is stored.
                                  [required; default: ./taxonomy_data; config:
                                  'train.train_args.data_path']
  --ckpt-output-dir PATH          output directory to store checkpoints in
                                  during training  [required; default:
                                  checkpoints; config:
                                  'train.train_args.ckpt_output_dir']
  --data-output-dir PATH          output directory to store training data in
                                  [required; default: train-output; config:
                                  'train.train_args.data_output_dir']
```

**Comment by cdoern** [comment]
Will give this a more thorough review this afternoon but let me just say this is FANTASTIC!!!! This is the exact direction I wanted to go in with the config parsing :raised_hands:

**Review by RobotSail — APPROVED** [review]
LGTM, if we can get this in before #1471

**Review by cdoern — APPROVED** [review]

**Pull request merged**